pynav/batch.py

```diff
@@ -0,0 +1,313 @@
+        self,
+        states: List[State],
+        compute_jacobians: List[bool] = None,
+    ) -> Tuple[np.ndarray, List[np.ndarray]]:
+        """A generic process residual.
+
+        The input self._u is used to propagate the state
+        at time x_km1 forward in time to genrate x_k_hat.
+        An error is created as 
+
+            e = x_k.minus(x_k_hat),
+        where x_k is our current operating point at time k.
+        """
+        x_km1 = states[0]
+        x_k = states[1]
+        dt = x_k.stamp - x_km1.stamp
+
+        # Evaluate the process model, compute the error
+        x_k_hat = self._process_model.evaluate(x_km1.copy(), self._u, dt)
+        e = x_k.minus(x_k_hat)
+
+        # Scale the error by the square root of the info matrix
+        L = self._process_model.sqrt_information(x_km1, self._u, dt)
+        e = L.T @ e
+
+        # Compute the Jacobians of the residual w.r.t x_km1 and x_k
+        if compute_jacobians:
+            jac_list = [None] * len(states)
+            if compute_jacobians[0]:
+                jac_list[0] = -L.T @ self._process_model.jacobian(
+                    x_km1, self._u, dt
+                )
+            if compute_jacobians[1]:
+                jac_list[1] = L.T @ x_k.minus_jacobian(x_k_hat)
+
+            return e, jac_list
+
+        return e
+
+
+class MeasurementResidual(Residual):
+    def __init__(self, keys: List[Hashable], measurement: Measurement):
+        super().__init__(keys)
+        self._y = measurement
+
+    def evaluate(
+        self,
+        states: List[State],
+        compute_jacobians: List[bool] = None,
+    ) -> Tuple[np.ndarray, List[np.ndarray]]:
+        """A generic residual for a measurement.
+
+        Computes a residual by comparing the true measurement
+        to the predicted measurement given by the model.
+
+        The Jacobian of the residual with respect to the state
+        is then the negative of the measurement model Jacobian.
+        """
+        # Extract state
+        x = states[0]
+
+        # Compute predicted measurement
+        y_check = self._y.model.evaluate(x)
+        e = self._y.value.reshape((-1, 1)) - y_check.reshape((-1, 1))
+
+        # Weight error by square root of information matrix
+        L = self._y.model.sqrt_information(x)
+        e = L.T @ e
+
+        if compute_jacobians:
+            jacobians = [None] * len(states)
+
+            if compute_jacobians[0]:
+                jacobians[0] = -L.T @ self._y.model.jacobian(x)
+            return e, jacobians
+
+        return e
+
+
+@dataclass
+class OptimizationSettings:
+    """A dataclass to hold some default optimizaiton settings."""
+    solver: str = "GN" # Solver type, either "GN" or "LM"
+    max_iters: int = 100 # maximum number of iters
+    step_tol: float = 1e-7 # Convergence tolerance, norm of the step size.
+    tau: float = 1e-11 # tau parameter in LM
+    verbose: bool = True # Print convergence during runtime 
+
+
+def run_batch(
+    x0: State,
```

Comment:
just a thought: can we add the prior as an optional argument? i think its possible to solve batch without it depending on the problem. not sure. maybe not a good idea